crates/prover/src/runner.rs

```diff
@@ -467,83 +491,34 @@ where
-                .expect("There should be a state root");
-            let initial_batch_hash = self
-                .ledger_db
-                .get_soft_confirmation_by_number(&BatchNumber(first_l2_height_of_l1))?
-                .ok_or(anyhow!(
-                    "Could not find soft confirmation at height {}",
-                    first_l2_height_of_l1
-                ))?
-                .prev_hash;
-
-            let final_state_root = self
-                .ledger_db
-                .get_l2_state_root::<Stf::StateRoot>(last_l2_height_of_l1)?
-                .expect("There should be a state root");
-
-            let (inclusion_proof, completeness_proof) = self
-                .da_service
-                .get_extraction_proof(l1_block, &da_data)
-                .await;
-
-            let transition_data: StateTransitionData<Stf::StateRoot, Stf::Witness, Da::Spec> =
-                StateTransitionData {
-                    initial_state_root,
-                    final_state_root,
-                    initial_batch_hash,
-                    da_data,
-                    da_block_header_of_commitments,
-                    inclusion_proof,
-                    completeness_proof,
-                    soft_confirmations,
-                    state_transition_witnesses,
-                    da_block_headers_of_soft_confirmations,
-                    sequencer_commitments_range: (
-                        0,
-                        (sequencer_commitments.len() - 1)
-                            .try_into()
-                            .expect("cant be more than 4 billion commitments in a da block; qed"),
-                    ), // for now process all commitments
-                    sequencer_public_key: self.sequencer_pub_key.clone(),
-                    sequencer_da_public_key: self.sequencer_da_pub_key.clone(),
-                };
-
-            let should_prove: bool = {
-                let mut rng = rand::thread_rng();
-                // if proof_sampling_number is 0, then we always prove and submit
-                // otherwise we submit and prove with a probability of 1/proof_sampling_number
-                if prover_config.proof_sampling_number == 0 {
-                    true
-                } else {
-                    rng.gen_range(0..prover_config.proof_sampling_number) == 0
-                }
-            };
-
-            // Skip submission until l1 height
-            if l1_height >= skip_submission_until_l1 && should_prove {
-                self.generate_and_submit_proof(transition_data, l1_height, hash)
+            let proving_session_exists = self
+                .check_and_recover_ongoing_proving_sessions(l1_height)
+                .await?;
+            if !proving_session_exists {
+                // There is no ongoing bonsai session to recover
+                let transition_data: StateTransitionData<Stf::StateRoot, Stf::Witness, Da::Spec> =
+                    self.create_state_transition_data(
```

Comment:
i don't think we should check this all the time, just on startup is fine, no?

Reply:
done